src/driver/amdxdna/ve2_hwctx.c

```diff
@@ -89,6 +239,326 @@ static int ve2_create_host_queue(struct amdxdna_dev *xdna, struct ve2_hsa_queue
+		struct amdxdna_gem_obj *abo;
+		void *cmd_data;
+		u32 cmd_data_len;
+
+		abo = amdxdna_gem_get_obj(hwctx->client, boh, AMDXDNA_BO_CMD);
+		if (!abo) {
+			XDNA_ERR(xdna, "Failed to find cmd BO %d", boh);
+			return -ENOENT;
+		}
+
+		cmd_data = amdxdna_cmd_get_payload(abo, &cmd_data_len);
+		if (!cmd_data) {
+			XDNA_ERR(xdna, "Invalid command data received");
+			amdxdna_gem_put_obj(abo);
+			return -EINVAL;
+		}
+
+		if (get_ve2_dpu_data_next(cmd_data))
+			ret = submit_command_indirect(hwctx, cmd_data, seq);
+		else
+			ret = submit_command(hwctx, cmd_data, seq);
+		if (ret) {
+			XDNA_ERR(xdna, "Submit chain command(%d) failed, error %d", i, ret);
+			amdxdna_gem_put_obj(abo);
+			return ret;
+		}
+
+		amdxdna_gem_put_obj(abo);
+	}
+
+	return ve2_hwctx_add_job(hwctx, job, *seq, cmd_chain->command_count);
+}
+
+int ve2_cmd_submit(struct amdxdna_ctx *hwctx, struct amdxdna_sched_job *job, u32 *syncobj_hdls,
+		   u64 *syncobj_points, u32 syncobj_cnt, u64 *seq)
+{
+	struct amdxdna_dev *xdna = hwctx->client->xdna;
+	struct amdxdna_gem_obj *cmd_bo = job->cmd_bo;
+	int ret;
+	u32 op;
+
+	op = amdxdna_cmd_get_op(cmd_bo);
+	if (op != ERT_START_DPU && op != ERT_CMD_CHAIN) {
+		XDNA_WARN(xdna, "Unsupported ERT cmd: %d received", op);
+		return -EINVAL;
+	}
+
+	if (op == ERT_CMD_CHAIN)
+		ret = ve2_submit_cmd_chain(hwctx, job, seq);
+	else
+		ret = ve2_submit_cmd_single(hwctx, job, seq);
+
+	if (ret) {
+		XDNA_ERR(xdna, "Failed to submit a command. ret %d\n", ret);
+		return ret;
+	}
+
+	ve2_ring_doorbell(hwctx);
+
+	return 0;
+}
+
+/*
+ * Handling interrupt notification based on read_index and write_index.
+ */
+#define check_read_index	\
+	({			\
+	static u64 counter;	\
+	u64 *read_index = (u64 *)((char *)priv_ctx->hwctx_hsa_queue.hsa_queue_p +	\
+		HSA_QUEUE_READ_INDEX_OFFSET);						\
+	if (counter % print_interval == 0) {						\
+		XDNA_DBG(xdna, "read idx addr (0x%llx)", (u64)read_index);		\
+		XDNA_WARN(xdna, "hwctx [%p] check read idx (%lld) > cmd idx (%lld)",	\
+			  hwctx, *read_index, seq);					\
+	}			\
+	counter++;		\
+	((*read_index) > seq);	\
+	})
+
+int ve2_cmd_wait(struct amdxdna_ctx *hwctx, u64 seq, u32 timeout)
+{
+	struct amdxdna_ctx_priv *priv_ctx = hwctx->priv;
+	struct ve2_hsa_queue *queue = &priv_ctx->hwctx_hsa_queue;
+	struct amdxdna_client *client = hwctx->client;
+	struct amdxdna_dev *xdna = client->xdna;
+	struct amdxdna_sched_job *job;
+	u32 print_interval = 300;
+	unsigned long wait_jifs;
+	int ret = 0;
+
+	/*
+	 * NOTE: this is simplified hwctx which has no col_entry list for different ctx
+	 * sharing the same lead col.
+	 * The current version assumes one hwctx is 1:1 mapping with one lead cert col
+	 */
+	wait_jifs = msecs_to_jiffies(timeout);
+	mutex_lock(&queue->hq_lock);
+	if (wait_jifs)
+		ret = wait_event_interruptible_timeout(priv_ctx->waitq, check_read_index,
+						       wait_jifs);
+	else
+		ret = wait_event_interruptible(priv_ctx->waitq, check_read_index);
+	mutex_unlock(&queue->hq_lock);
```

Comment:
Wait event while holding a mutex lock is a bad idea. Can we avoid this kind of design?